CFE-815: Route secret injection for external certificate management (#1307)
openshift/enhancements: enhancements/ingress/route-secret-injection-for-external-certificate-management.md

integration with third-party certificate management solutions like cert-manager and
avoid manual certificate management by the user which is more error prone.

## Motivation

OpenShift customers currently manually manage certificates for user workloads
by updating OpenShift Routes with the updated certificate data during expiry/renew
workflow. This is cumbersome activity if users have a huge number of workloads and
is also error prone.

This enhancement adds the support to OpenShift Routes for third-party certificate
management solutions like cert-manager by extending the Route API to read the serving
certificate data via a secret reference.

### User Stories

- As an end user of Route API, I want to be able to provide a tls secret reference on
  OpenShift Routes so that I can integrate with third-party certificate management solution

- As an OpenShift cluster administrator, I want to use third-party solutions like cert-manager
  for certificate management of user workloads on OpenShift so that no manual process is
  required to renew expired certificates.

- As an Openshift engineer, I want to be update the router so that it is able read secrets directly

>   if all the preconditions have been met by the router serviceaccount.

Author: Addressed.

reference field described under [API Extensions](#api-extensions).

- The end user must have generated the serving certificate generated
  as a pre requisite using third-party systems like cert-manager.
- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources)
  CR must be created in the same namespace where the Route is going to be created.
- The end user must create a role and in the same namespace as the secret containing the certificate from earlier,
  ```bash
  oc create role secret-reader --verb=get,list,watch --resource=secrets --resourceName=<secret-name>
  ```
- The end user must create a rolebinding in the same namespace as the secret
  and bind the router serviceaccount to the above created role.
  ```bash
  oc create rolebinding foo-secret-reader --role=secret-reader --serviceaccount=openshift-ingress:router --namespace=<current-namespace>
  ```
- To expose a user workload, the user would create a new Route with the
  `.spec.tls.externalCertificate` referencing the generated secret that was created
  in the previous step.
- If the secret that is referenced exists and has a successfully generated

>   cert/key pair, the router will serve this certificate if all preconditions are met.

Author: Linked to Implementation section.

N.A

### API Extensions

A `.spec.tls.externalCertificate` field is added to Route `.spec.tls` which can be used to provide a secret name
containing the certificate data instead of using `.spec.tls.certificate` and `spec.tls.key`.

```go

// TLSConfig defines config used to secure a route and provide termination
//
// +kubebuilder:validation:XValidation:rule="has(self.termination) && has(self.insecureEdgeTerminationPolicy) ? !((self.termination=='passthrough') && (self.insecureEdgeTerminationPolicy=='Allow')) : true", message="cannot have both spec.tls.termination: passthrough and spec.tls.insecureEdgeTerminationPolicy: Allow"
// +openshift:validation:FeatureSetAwareXValidation:featureSet=TechPreviewNoUpgrade;CustomNoUpgrade,rule="!(has(self.certificate) && has(self.externalCertificate))", message="cannot have both spec.tls.certificate and spec.tls.externalCertificate"
type TLSConfig struct {
	// ...

	// externalCertificate provides certificate contents as a secret reference.
	// This should be a single serving certificate, not a certificate
	// chain. Do not include a CA certificate. The secret referenced should
	// be present in the same namespace as that of the Route.

> 	// Forbidden when `certificate` is set.

Reviewer: Should we mention the fact that the RBAC needs to be done by the user?

Author: We could do that. But I don't see a concise way to mention this detail. Either we'll only have to mention, this field requires manual rbac to be created and maybe link to docs. Putting the entire rbac (role, rolebinding, etc) would be too bloated I feel. Moreover there are couple of variations in rbac as well. 
```
- Grant router service account access to secret by secret-name (explicit rbac)
- Grant router service account access to all secrets in a fixed namespace (implicit rbac)
```

Reviewer: I thought of a simple phrase without any doc references which can give a motivation to the further digging though. Something like: "Router service account needs to be granted with read-only access to this secret."

Author: I've added a note in `TP -> GA criteria` to update the API. Since I didn't want to start a new API update cycle. Hope this is fine.

Reviewer: Yes, it's fine with me. I think we cannot have enough docs highlighting this fact. I'm sure the OCP docs will be on point but I believe that `oc explain` should be ready too.

	//
	// +openshift:enable:FeatureSets=CustomNoUpgrade;TechPreviewNoUpgrade
	// +optional
	ExternalCertificate LocalObjectReference `json:"externalCertificate,omitempty" protobuf:"bytes,7,opt,name=externalCertificate"`
}

// LocalObjectReference contains enough information to let you locate the
// referenced object inside the same namespace.
// +structType=atomic
type LocalObjectReference struct {
	// name of the referent.
	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
	// +optional
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
}
```

> _Note_: The default value would be `nil`. The secret is required to be created

Author: Correct, missed this. The field was a pointer before. Removed this statement.

in the same namespace as that of the Route. The secret must be of type
`kubernetes.io/tls` and the tls.key and the tls.crt key must be provided in
the `data` (or `stringData`) field of the Secret configuration.

If neither `.spec.tls.externalCertificate` or `.spec.tls.certificate` and `.spec.tls.key` are

> provided the router will serve the default generated certificates.

Author: User's will not be able to provide both `.spec.tls.certificate/key` and `.spec.tls.externalCerificate`. API server admission validation will enforce this.

All valid and invalid scenarios will be depicted via the existing `RouteIngressCondition`.

#### Variation

N.A

### Implementation Details/Notes/Constraints [optional]

The router will read the secret referenced in `.spec.tls.externalCertificate` if present and

> if the following pre-conditions (validated in the API server) are met it uses

Author: Re-arranged this section. Should be fixed now. The conditions/validation criteria is what I was referring to here.

> if the following pre-conditions (validated in the API server) are met it uses

Reviewer: Ah, did you mean the 2 validations listed below as "the following preconditions"? May be a rephrase similar to something like this will prevent the misinterpretations:

> The router will read the secret referenced in `.spec.tls.externalCertificate` and will use the certificate inside to configure HAProxy if the secret is present and if the following pre-conditions are met:

this certificate us configure haproxy.

> The router will bootstrap a watch based secret manager to ensure it can keep the

Author: Something I plan to add ([ref](https://github.com/openshift/router/pull/482/files#diff-6609f44f0a0cf86b0a99e92cfcf047039b6a42e64dc45d502234743b56410678))

> The router will bootstrap a watch based secret manager to ensure it can keep the

Reviewer: I see. Then it deserves some prior description because the reader doesn't know anything about it.

> The router will bootstrap a watch based secret manager to ensure it can keep the

Author: Added a new section mentioning this component.

certificate/secret up-to-date. This means the router pod will maintain active watches
for every secret that is referenced by a route.

Every active watch will be linked to a route, meaning the watch based secret manager
will be linked to the lifecycle of the route. For every new route that is created,
the secret manager will start a watch if the route uses `.spec.tls.externalCertificate`.

> For every update route event, the secret manager only increments the reference count.

Author: The statement was incorrect, fixed it.

> For every update route event, the secret manager only increments the reference count.

Author: This part is the implementation level detail of the the secret manager/monitor which may or may not end up in o/library-go. So didn't want to add too many details here.

> For every update route event, the secret manager only increments the reference count.

Reviewer: We can edit the EP post implementation. For the moment, it's hard to understand the purpose of this reference counter so it needs a better description.

- Verify certificate and key (PEM encode/decode)
- Verify private key matches public certificate

Validations done by API server as part of [ValidateRoute()](https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go#L21),

- The secret created should be in the same namespace as that of the route.
- The secret created is of type `kubernetes.io/tls`.
- The router serviceaccount must have permission to read this secret particular secret.
  - The role and rolebinding to provide this access must be provided by the user.
- CEL validations will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
  are not specified on the route.

### Risks and Mitigations

There is a possibility of an invalid route being processed by the router (edge case),
if any changes are done to the rbac or the referenced secret is deleted after the API
server validation but before router has processed the route (maybe router pod is not running)
then this can lead to the router processing this incorrect route.

> > Will need to duplicate the validations present on the API server to the router.

Reviewer: Which doesn't eliminate the race, it just shortens the race window. Is it worth the effort? If the rbac or the secret is missing at the time the router needs them, I suppose the route's status will be updated. I think that's the best effort we can do in such a unfortunate scenario.

Reviewer: I just realized that the route's status won't be updated with the missing RBAC condition because if the route is created than it was accepted by the api server. I agree with this statement then, we have to check the same in the router.

Reviewer: We can resolve this conversation.

### Drawbacks

The user will need to manually create, provide and maintain the rbac required by the
router so that it can access secrets securely. This becomes tedious when users have

> 100s of Routes.

Reviewer: ```suggestion
many Routes.
```

- Updating routes from default certificates to certificate referenced via
  secrets and vice-versa.
- Updating secret/certificate referenced in routes and verify serving
  certificate has been updated.
- Updating secret/certificate with incorrect information and verify route
  is not admitted due to validation failure. (eg: mismatched public and private key, etc)

### Graduation Criteria

This feature will initially be released as Tech Preview only. The e2e tests
in openshift/origin will only be added when graduating this feature to GA.

#### Dev Preview -> Tech Preview

N/A. This feature will go directly to Tech Preview.

#### Tech Preview -> GA (Future work)

The e2e tests as part of openshift/origin should be consistently passing.
The router will need to undergo performance testing as part of OCP payload
to ensure the memory implications of creating and maintains all the active watches
is verified to be efficient.

##### Future work

The ingress-to-route controller in the route-controller-manager will need to
be updated to ensure that the created routes use `.spec.tls.externalCertificate`
instead of `.spec.tls.certificate`. Additional tests will need to be added into

> o/origin for this scenario.

Reviewer: That's more a question for the future but who will do the necessary RBAC in that case?

Author: I think we can still rely on the user providing this rbac if we decide to implement this in the future.

Reviewer: Is it only me reading this like `.spec.tls.externalCertificate` will be preferred solution over `.spec.tls.certificate` when the former will become GA?

Author: That is what I had in mind, but we can take a collective call. Figured in the future when KMS is brought in it is better if we have the certificate in a secret as opposed to in a route.

- Provide smooth roll out of new certificates on OpenShift router when referenced certificates
  are renewed (secret containing the certificate is updated).

### Non-Goals

- Provide certificate life cycle management controls on the Route API (expiryAfter, renewBefore, etc).
- Modify ingress-to-route controller behaviour to use `.spec.tls.certificateRef`
- Extend this feature to cover CA certificate or destination CA certificate in the Route API.

## Proposal

This enhancement proposes extending the openshift/router to read serving certificate data either
from the Route `.spec.tls.certificate` and `.spec.tls.key` or from a new field `.spec.tls.certificateRef`
which is a `kubernetes.io/tls` type secret reference. This `certificateRef` field will enables the
users to provide a reference to a Secret containing the serving cert/key pair that will be parsed
and served by OpenShift router.

### Workflow Description

The following workflow describes the integration with third party
certificate management solutions like cert-manager with the certificate
reference field described under [API Extensions](#api-extensions).

- The end user must have generated the serving certificate generated

>   as a pre requisite using third-party systems like cert-manager.

Reviewer 2: *nit*
```suggestion
  as a prerequisite using third-party systems like cert-manager.
```

---
title: route-secret-injection-for-external-certificate-management
authors:
  - "@thejasn"
reviewers:
  - "@Miciah"
  - "@alebedev87"
  - "@tgeer" 
  - "@joelspeed"
  - "@deads2k" 
approvers:
  - "@Miciah" 
api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
  - "@joelspeed" 

> creation-date: 20222-12-13

Reviewer: ```suggestion
creation-date: 2022-12-13
```

> last-updated: 20223-01-10 

Reviewer: ```suggestion
last-updated: 2023-01-10 
```

tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
  - https://issues.redhat.com/browse/CFE-704
see-also:
  - "/enhancements/this-other-neat-thing.md"
replaces:
  - "/enhancements/that-less-than-great-idea.md"
superseded-by:

>   - "/enhancements/our-past-effort.md"

Reviewer: Do we need to keep them if they are of no use?

Author: Addressed.

with third-party certificate management solutions like [cert-manager](https://github.com/cert-manager/cert-manager).
And this is mainly due to the restriction on Routes API requiring the certificate
data to be burned in the CR as opposed to having a reference. This is especially
problematic when third-party solutions also manage the life cycle (create/renew/delete)
of the generated certificates which OpenShift Routes does not support and requires
manual intervention to manage certificate life cycle.

This enhancement aims to provide a solution where OpenShift Routes can support
integration with third-party certificate management solutions like cert-manager and
avoid manual certificate management by the user which is more error prone.

## Motivation

OpenShift customers currently manually manage certificates for user workloads
by updating OpenShift Routes with the updated certificate data during expiry/renew
workflow. This is cumbersome activity if users have a huge number of workloads and
is also error prone. 

This enhancement adds better support to OpenShift Routes for third-party certificate
management solutions like cert-manager.

> ### User Stories

Author: I've added a few additional stories to cover both the end user and engineers requirements. I hope this makes it clear. The first few stories are mainly from the end user's perspective (so kind of like goals), where as the last three stories are more along the work required.

> ### User Stories

Reviewer: Seems like I was misinterpreting the "User Stories" chapter. It's [supposed to describe the use cases](https://github.com/openshift/enhancements/blob/master/guidelines/enhancement_template.md#user-stories) which would be enabled after the implementation of the EP.

* As an OpenShift cluster administrator, I want to be able to provide a certificate reference on 

> OpenShift Routes so that I can integrate with third-party certificate management solution

Author: I've updated the actor for some of the user stories, except the second one as this kind of hints towards installing third-party operator.

* As an OpenShift cluster administrator, I want to use third-party solutions like cert-manager
for certificate management of user workloads on OpenShift so that no manual process is
required to renew expired certificates.

* As an OpenShift cluster administrator, I want OpenShift Routes to support both manual/managed
mode of operation for certification management so that I can switch between manual certificate
management and third-party certificate management.

* As an OpenShift engineer, I want to be able to update Route API so that I can integrate
OpenShift Routes with third-party certificate management solutions like cert-manager.

### Goals

- Provide a user configurable option in Route API to provide externally managed certificate data.
- Provide users a mechanism to switch between using externally managed certificates and
  manually managing certificates on OpenShift routes.
- Provide smooth roll out of new certificates on OpenShift router when managed certificates
  are renewed.

> ### Non-Goals

Author: I've added this case case well.

> ### Non-Goals

Reviewer: Sorry I didn't explain myself well and I mixed up the chapters too. What I wanted to say is that if the secret referenced in the annotation and injected by the new controller gets removed - we wipe out the certificate data from the route. This is a goal (not a non-goal, sorry) - certificate data cleanup.

> ### Non-Goals

Author: Addressed. Now covered as part of 
> Provide users with a mechanism to switch between using externally managed certificates and
  manually managed certificates on OpenShift routes and vice-versa.

- Provide certificate life cycle management controls on the Route API (expiryAfter, renewBefore, etc).
- Provide migration of certificates from current solution.

## Proposal

> This enhancement proposes introducing a new controller (secret-injector) in route-controller-manager

Reviewer: Can you please add a link to the route controller manager repository or doc?

Btw, about the naming, strictly speaking the new controller will inject a certificate not secret.

Author: Addressed. As per discussion, controller name will be retained as `secret-injector`.

Reviewer: For the record: the reason to stick with `secret-injector` was the consistency with the route annotation which mentions the secret.

> to manage a new annotation on the Route CRD. This annotation enables the users to provide a reference to 

Reviewer: Annotation won't need a change in the CRD. We can say something like `on the Route object`.

Author: Addressed.

> a Secret containing the serving cert/key pair that will be served by OpenShift router.

Reviewer: I think it's worth mentioning  mentioning that the contents of the secret will be added to `.spec.tls` fields.

Author: Addressed.

This annotation will be given a higher preference if route CR also has `.spec.tls.Certificate` 
and `.spec.tls.Key` defined.

### Workflow Description

> End users will have 2 possible variations when using this functionality,

Author: Addressed.

- Create an ingress for the user workload and a managed route is created automatically
  by the ingress-to-route controller.
- Create a route for the user workload and this is completely managed by the end user.

Both these workflows will support integrating with third party certificate management
solutions like cert-manager with the annotation described below.

**When an Ingress is used to expose user workload**
- The end user must have generated the serving certificate as a pre requisite
  using third-party systems like cert-manager.

> - In cert-manager's case, the Certificate CR must be created in the same namespace

Author: Addressed

  where the Ingress is going to be created.
- To expose a user workload, he would need to create a new Ingress with the generated secret
  referenced in `.spec.tls.secretName`.
- If the secret CR that is referenced exists and has a successfully generated
  cert/key pair generated, the ingress-to-route controller copies this certificate data 
  into the route at `.spec.tls.certificate` and `.spec.tls.key`.

**When a Route is directly used to expose user workload**
- The end user must have generated the serving certificate as a pre requisite
  using third-party systems like cert-manager.
- To expose a user workload, the user would create a new Route with an annotation
  referencing the generated secret that was created in the previous step.
- If the secret that is referenced exists and has a successfully generated
  cert/key pair generated, this certificate data is copied into the route at 

>   `.spec.tls.certificate` and `.spec.tls.key`.

Author: Addressed

*Note*: The same workflow would apply to updating an existing Route to integrate
with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate` 
and `.spec.tls.key` in the route CR since the annotation gets a higher preference.

#### Certificate Life cycle

Post integrating with third-party solutions like cert-manager, the end user can also
disable automatic certificate management. This can be done in 2 ways,
- The end user can delete the annotation on the Route CR.
- The end user deletes the secret that was associated with the Certificate CR.

> **Deleting the annotation on the Route CR**

Author: Addressed. Route CR -> Route throughout the EP.

The end user can delete the annotation on the Route CR, this will result
in abondoning the cert/key pair that was generated via cert-manager and resort to
using the default certificates that were initially generated by the cluster-ingress-operator
i.e. clearing the `.spec.tls` on Route. 

**Deleting the generated secret that contains the serving cert/key pair**
When the end user deletes the secret containing the certificate data, the `.spec.tls` on 
Route will be cleared. This will result in the route serving default certificates. In such

> a scenario the annotation that was added by the user will be automatically deleted by the 

Author: >The deletion of the annotation seems a little odd to me. The user has expressed its intent by adding the annotation. The intent may stay the same even if the secret reference is not there any longer. Comparing this to the Ingress case: if the secret is removed would Ingress' .spec.tls.secretName field be wiped too? I don't think so.

As for the ingress case, if the `secretName` is filed and the secret is not found the reconciliation is skipped and waits  until the secret is present (so no Route is created or the route gets deleted). And if the field is not set, it pivots to using default certs. We can't do a similar thing here since the Route is created by the user and we can skip reconciling new routes, but for updates the implementation will have gaps. Hence why I decided to clear the field since we are talking about managed certificates and the controller manages `.spec.tls` when the annotation is present. 

>Also, I'm wondering how the new controller would find out that the secret is deleted and link it back to the route whose certificate needs to be cleaned. A watch on the secrets? That would need something in the secret object to link it back to the route (annotation, status etc.).

This gives a little more idea into the implementation https://github.com/openshift/enhancements/pull/1307#discussion_r1072088428.

>Overall I think the secret removal needs to be a little more detailed in the workflow chapter.
I'll add a bit more context on this.

> a scenario the annotation that was added by the user will be automatically deleted by the 

Author: Updated with the above details.

secret-injector controller. This is only in the case where the user is directly creating 
routes instead of relying on the Ingresses.

#### Variation [optional]

N.A

### API Extensions

A new annotation of type `string` to be introduced as part of the integration
to support third-party certificate management systems,

```yaml
annotations:

>   route.openshift.io/tls-secret-reference: <secret-name>

Reviewer: I'm thinking whether `reference` is not too generic as a name. The reference can be a namespaced reference too while here we assume that the secret is from the same namespace. What do you think about `tls-secret-name`?

Author: Renamed to `tls-secret-name`.

```

> *Note*: The default value would be N/A.

Reviewer: The expected contents of the secret as well as  the assumed namespace should be described too.

Author: Addressed

This annotation can be applied by the end user on Route CR. The controller that 

> will be introduced as part of this enhancement will be responsible in
> process this annotation on the Route CR.

Reviewer: ```suggestion
will be introduced as part of this enhancement will be responsible for the 
processing of this annotation on the Route CR.
```

The Route API will also be updated to denote certificate status under `.status`,

```go

// These are valid conditions of pod.
const (

    // ....
    
    // RouteCertificateDefaultType denotes that the user has not provided 
    // any custom certificate and the default generated certificate is being
    // served on the route.
    RouteCertificateDefaultType RouteIngressConditionType = "Default"

>     // RouteCertificateCustomType denotea that there is a custom certificate 

Reviewer: ```suggestion
    // RouteCertificateCustomType denotes that there is a custom certificate 
```

    // being served on the route. 
    RouteCertificateCustomType RouteIngressConditionType = "Custom"

    // RouteCertificateManagedType denotes that the annotation is in use to 
    // inject a third-party managed secret containing the certificate that 
    // is being served on the route.
    RouteCertificateManagedType RouteIngressConditionType = "Managed"

)

```

#### Variation

##### Alternative to the tls-secret-reference annotation

***Note***: This idea has been dropped in favor of a secret reference through an annotation.

The alternative to the annotation was to provide a new API field in the Route API,
through which the user could provide the certificate reference.

```go
type CertificateReference struct {
	// certificate resource name.
	// +optional
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
}

Type RouteSpec struct {
	// ...

	CertificateRef CertificateReference `json:"certificateRef,omitempty" protobuf:"bytes,5,opt,name=certificateRef"`
}
```

The reasoning for the field was mainly for future proofing the API to handle integrations

> with other third-party certificate management systems. But this API introduces inconsistency

Reviewer: Can you please elaborate on which inconsistency turned us way from this alternative?

Author: Addressed

and does not fare well when the user wants to Ingress over directly creating routes.

#### Alternative to new status condition types

An alternative to introducing 3 new condition types, would be to use `metadata.managedFields` 
on Route to denote to the user which fields are managed by the secret-injector controller and 
server-side apply would be used by the controller during updates to ensure that `.spec.tls.certificate` 
and `.spec.tls.secret` are always managed by the controller as long as the annotation is present.

This although helps in avoiding addition of new condition types will result in a few inconsistencies,

- When transitioning from using managed certificates to manually managing custom certificates,
  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update 
  operation done by the user is only on the `.metadata.annotations`. 
- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.

### Implementation Details/Notes/Constraints [optional]

The new controller (secret-injector) will be responsible for watching routes and processing those

> that have the valid annotation for `tls-secret-reference`. This controller will also update
> `.status` of the route to contain the latest information on the certificate in use.

Reviewer: ```suggestion
that have the valid annotation for `tls-secret-reference`. This controller will also update
`.status` of the route to signal the latest updates on the certificate in use.
```

In scenarios where the user deletes the annotation, it is the controller's 

> responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e. 

Reviewer: Assuming that the new controller will be watching the routes using a predicate on the new annotation. Will the controller receive an event when this annotation is removed?

Author: The new controller will watch all routes and process all routes in order to update `.status.certificateCondiitions`. The ones with the annotation will have additional logic to update `.spec.tls`. 

The only filter on the route informer is going to be excluding those with ingress owner references.

Reviewer: I see, that's quite an overhead as we could use the new annotation as a predicate which (I guess) would enable the server side filtering instead of the client side (new controller). However that's maybe the only way around to clear `.spec.tls` fields after the annotation is removed.

Can you state this implementation detail in the EP please?

Author: Updated with the details.

if the status denotes `RouteCertificateManagedType` and since the required annotation is 
not present this results in the `.spec.tls` being cleared out and the `.status` updated 
to `RouteCertificateDefaultType`.

If the user deletes the secret associated with the Route through the `route.openshift.io/tls-secret-reference`

> annotation, the secret-injector controller will delete this annotation automatically and publish 

Reviewer: Same question as in the workflow section: how exactly the new controller will know about the secret removal? A dedicated watch? How to link the secret back to the route?

Author: We will have a secret informer, that will trigger a re-sync of all routes in the same namespace as that of the secret. This logic is very similar to what is already present in route-controller-manager ([ref](https://github.com/openshift/route-controller-manager/blob/4d25295dbf9821107fa9adeb968220877b3331e6/pkg/route/ingress/ingress.go#L205)).

Reviewer: it's sub-optimal but would do indeed.

Author: Updated with the details.

an `Event` so that the end user is notified regarding this switch to using the default generated
certificates since the associated secret is not found. The `.status` is also updated to 
`RouteCertificateDefaultType`.

When transitioning from using a managed certificate to manually managed certificate, the user
is expected to only delete the annotation. If the previously used `.spec.tls` is not deleted
by the user the Route will continue to use the same certificate data and `.status` is updated 
to `RouteCertificateCustomType`.

The new secret-injector controller will not reconcile Route CRs that are owned by the 
ingress-to-route controller.

### Risks and Mitigations

The new controller that is going to be introduced for this enhancement will have the capability 

> to read secret objects and write the secret data into the route CR. There is precedence 

Reviewer: is it really a risk if the `route-controller-manager` already has those rights?

Author: I have mentioned this here as the Auth team whom we took over cert-manager-operator from were initially told by the Networking team that using secrets for certificates was a no go as this increases the attack surface. But as it turns out,
> There is precedence already for such intrusive permissions required by the ingress-to-route controller.

So it is kind of mitigated.

Author: I'm open to removing this as a risk.

Reviewer: Using secrets on the edge workloads such as router is no go. Here we are inside the cluster and `route-controller-manager` already has those rights. So, yes I think those rights can be considered as given.

Author: Removed content under Risks.

### Drawbacks

This introduces an inconsistency between Ingress and Route CRD with respect to how 
a secret reference can be provided with the certificate data. The ingress has a field
`.spec.tls.secretName` where as the Route will have an annotation.

## Design Details

### Open Questions [optional]

1. Does this warrant making changes to the ingress-to-route controller in how 
   `.spec.tls.secretName` is processed? Currently, the controller reads the secret
   data and copies it over to the Route CR that is created. Since this enhancement
   introduces a difference between managed certificate provided via secrets and  
   manually provided secrets, this distinction isn't possible with Routes created
   via Ingresses.

   - Proposed change
     The ingress-to-route controller will also process the same annotation and if 
     present copies it over to the route that is created. This annotation will take
     precedence over `.spec.tls.secretName` and this offers enough distinction between
     third-party/manual certificate management.

**Answer**: The behaviour of the ingress-to-route controller will be retained as is and 

> no changes will be done.

Reviewer: Would be nice to document the reason why.

Author: Addressed.

operators e2e test suite. 

1. Test Route API without interfacing with Ingresses

    a. Create a edge terminated route that is using default certificates. 
    b. Update route with the new annotation and ensure the `.spec.tls` and `.status`
       is updated.
    c. Verify that the associated openshift routers are also serving the updated
       certificates and not the default certificates.
    d. Update certificate data and verify if the certificates served by the router 
       are updated as well.
    d. Delete the annotation and ensure that the route `.spec.tls` and `.status` 
       denotes the usage of default certificates. Also verify the associated
       openshift router is serving the default certificates.  

Various other transitions, `Custom`->`Managed`-`Custom` will also be tested.

### Graduation Criteria

This is a user facing change and will directly go to GA. This feature requires an
update to Openshift Docs.

#### Dev Preview -> Tech Preview

N/A.  This feature will go directly to GA.

#### Tech Preview -> GA

N/A.  This feature will go directly to GA.

#### Removing a deprecated feature

N/A.

### Upgrade / Downgrade Strategy

> On downgrades, all routes using this annotation will continue to use the custom certificates

Reviewer: ```suggestion
On downgrades, all routes using `route.openshift.io/tls-secret-reference` annotation will continue to use the custom certificates
```

is also error prone.

This enhancement adds the support to OpenShift Routes for third-party certificate
management solutions like cert-manager by extending the Route API to read the serving
certificate data via a secret reference.

### User Stories

- As an end user of Route API, I want to be able to provide a tls secret reference on
  OpenShift Routes so that I can integrate with third-party certificate management solution

- As an OpenShift cluster administrator, I want to use third-party solutions like cert-manager
  for certificate management of user workloads on OpenShift so that no manual process is
  required to renew expired certificates.

- As an Openshift engineer, I want to update the router so that it is able read
  secrets directly if all the preconditions have been met by the router serviceaccount.

  - The router serviceaccount must have permission to read this secret particular secret.
  - The role and rolebinding to provide this access must be provided by the user.

- As an OpenShift engineer, I want to update the route validation in openshift/library-go
  in order to validate the updated Route API.

>   - Both Openshift and Microshift both run the o/library-go validations as part of admission plugin

Reviewer: ```suggestion
  - Both Openshift and Microshift run the openshift/library-go validations as part of admission plugin
```

- As an OpenShift engineer, I want to be able to update Route API so that I can integrate
  OpenShift Routes with third-party certificate management solutions like cert-manager.

- As an OpenShift engineer, I want to be able to run e2e tests as part of openshift/origin
  so that testcases are executed to signal feature health by CI executions.

### Goals

- Provide users with a configurable option in Route API to reference externally managed certificates via secrets.
- Provide smooth roll out of new certificates on OpenShift router when referenced certificates
  are renewed (secret containing the certificate is updated).

### Non-Goals

- Provide certificate life cycle management controls on the Route API (expiryAfter, renewBefore, etc).

> - Modify ingress-to-route controller behaviour to use `.spec.tls.externalCertificate`

Reviewer: ```suggestion
- Modify ingress-to-route controller behaviour to use the external managed certificate reference from RouteAPI.
```
At this point the user doesn't know anything about the new API field.

- Extend this feature to cover CA certificate or destination CA certificate in the Route API.

## Proposal

This enhancement proposes extending the openshift/router to read serving certificate data either
from the Route `.spec.tls.certificate` and `.spec.tls.key` or from a new field `.spec.tls.externalCertificate`

> which is a `kubernetes.io/tls` type secret reference. This `externalCertificate` field will enables the
> users to provide a reference to a Secret containing the serving cert/key pair that will be parsed

Reviewer: ```suggestion
which is a `kubernetes.io/tls` type secret reference. This `externalCertificate` field will enable the
users to provide a reference to a secret containing the serving cert/key pair that will be parsed
```

The router will read the secret referenced in `.spec.tls.externalCertificate` and will use
the certificate inside to configure HAProxy if the secret is present and if the
following pre-conditions are met:

- Validations done by the router as part of [ExtendedValidateRoute()](https://github.com/openshift/router/blob/c407ebbc5d8d85daea2ef2d1ba539444a06f4d25/pkg/router/routeapihelpers/validation.go#L158) (contents of secret),

  - The secret created should be in the same namespace as that of the route.
  - The secret created is of type `kubernetes.io/tls`.
  - Verify certificate and key (PEM encode/decode)
  - Verify private key matches public certificate

- Validations done by API server as part of [ValidateRoute()](https://github.com/openshift/openshift-apiserver/blob/aac3dd5bf0547e928103a0f718ca104b1bb13930/pkg/route/apis/route/validation/validation.go#L21),

  - The router serviceaccount must have permission to read this secret particular secret.
    - The role and rolebinding to provide this access must be provided by the user.
  - CEL validations and o/library-go will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`

>     are not specified on the route.

Reviewer: This is like the first level validation which will be followed up by the extended validation in the router. So, I think this point can go first, before the router validation just to logically match the rel life workflow.

Author: Re-ordered the validations.

>   - CEL validations and o/library-go will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
>     are not specified on the route.

Reviewer: ```suggestion
  - CEL validations and openshift/library-go will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
    are not specified on the route at the same time.
```

The router being an edge component, from a security standpoint is more prone to
being compromised. In order to avoid providing the router with escalated privileges
to read all secrets, the router will implement a single item list/watch for secrets (secret monitor).
This uses name-scoped rbac (created by the user) to access the particular secrets.

A watch based secret monitor will be introduced in the router in order to keep
track of all the secrets referenced by the routes. This component is solely
responsible for maintaining all the single item list-watch functions required
to cache the referenced secrets.

The router will bootstrap the secret monitor to ensure it can keep the

> certificate/secret up-to-date. This means the router pod will maintain active watches

Reviewer: ```suggestion
certificate up-to-date. This means the router pod will maintain active watches
```
I suppose that the sectet will be kept up to date by the external certificate manager. The service monitor will solely read the secret and maintain the HAProxy certificate up to date.

for every secret that is referenced by a route.

Every active watch will be linked to a route, meaning the secret monitor
will be linked to the lifecycle of the route. For every new route that is created,
the secret monitor will start a watch if the route uses `.spec.tls.externalCertificate`.

> If a route is deleted, the secret monitor will deregister the route and teardown

Reviewer: Deregister the route? What does it mean exactly?

Author: Updated to be more generic and not use implementation specific function names.

the watch associated with it.

> The `ServiceAliasConfig` creation logic will be updated in the router to also parse

Reviewer: Would be good to reference `ServiceAliasConfig`'s code. For now this type comes a little bit out of a blue.

Author: Added a link.

the secret referenced in `.spec.tls.externalCertificate`. The router will
use the default certificates only when `.spec.tls.certificate/key` or `.spec.tls.externalCertificate`
are not provided.

The cluster-ingress-operator will propagate the relevant Tech-Preview feature gate down to the
router. This feature gate will be added as a command-line argument called `ROUTER_EXTERNAL_CERTIFICATE`
to the router and will not be user configurable.

### Risks and Mitigations

There is a possibility of an invalid route being processed by the router (edge case),
if any changes are done to the rbac or the referenced secret is deleted after the API
server validation but before router has processed the route (maybe router pod is not running)
then this can lead to the router processing this incorrect route.

> Will need to duplicate the validations present on the API server to the router.

### Drawbacks

> The user will need to manually create, provide and maintain the rbac required by the
> router so that it can access secrets securely. This becomes tedious when users have
> many of Routes.

Reviewer: ```suggestion
The user will need to manually create, provide and maintain the rbac required by the
router so that it can access secrets. This becomes tedious when users have
many routes.
```

The workaround for this is to document various levels of rbac that can be provided,

- Grant router service account access to secret by secret-name (explicit rbac)
- Grant router service account access to all secrets in a fixed namespace (implicit rbac)

The above variations need to be documented for the end user as part of OpenShift documentation.

## Design Details

### Open Questions [optional]

- Performance testing of openshift-router in tech-preview? Is there
  a workflow present where we can gather some early metrics (memory, cpu)? This will help in
  preemptively addressing performance concerns before going GA.

  > Will be addressed when taking feature from TP -> GA.

- Do we make changes to the ingress-to-route controller as well?

  > The ingress-to-route behaviour will remain as is i.e. it will not make use of
  > the newly introduced field.

> - What should be the behaviour of `ValidationHostUpdate()` when using `externalCertificate`?

Reviewer: Would be good to reference `ValidateHostUpdate` function like you did in the implementation details.

Author: Added a link.

with new tests utilizing `.spec.tls.externalCertificate`. Ensure the tests cover the following scenarios,

- Updating routes from default certificates to certificate referenced via
  secrets and vice-versa.
- Updating secret/certificate referenced in routes and verify serving
  certificate has been updated.
- Updating secret/certificate with incorrect information and verify route
  is not admitted due to validation failure. (eg: mismatched public and private key, etc)

### Graduation Criteria

This feature will initially be released as Tech Preview only. The e2e tests
in openshift/origin will only be added when graduating this feature to GA.

#### Dev Preview -> Tech Preview

N/A. This feature will go directly to Tech Preview.

#### Tech Preview -> GA (Future work)

The e2e tests as part of openshift/origin should be consistently passing.
The router will need to undergo performance testing as part of OCP payload
to ensure the memory implications of creating and maintains all the active watches
is verified to be efficient.

Update API godoc to document that manual intervention is required for using

> `.spec.tls.externalCertificate`.

Reviewer: You mean in this EP, additionally to what you already did? Or in the code? The latter should be part of DefintionOfDone during the implementation.

Author: It's the former, related to our previous conversation from https://github.com/openshift/enhancements/pull/1307#discussion_r1252571127. I've updated this to be a bit more specific so as to not forget.

On downgrades, all routes specifying `.spec.tls.externalCertificate` will switch over to use the default certificates
unless the route is manually edited and the `.spec.tls` is updated.

Upgrade strategy not considered since this feature is going to be added as TechPreviewNoUpgrade.

### Version Skew Strategy

This feature will be added as TechPreviewNoUprade.

### Operational Aspects of API Extensions

Route validation in the API server will be modified to validate the following scenarios,

- Check if secret/certificate referenced under `.spec.tls.externalCertificate` exists.
- Check if secret/certificate referenced under `.spec.tls.externalCertificate` is of the correct type.
- Check if router service account has permissions to read referenced secret.
- Check if route has only one of the fields set,
  - `.spec.tls.certificate` and `.spec.tls.key`
  - `.spec.tls.externalCertificate`

#### Failure Modes

##### Referenced secret not present

In scenarios where the secret has not been created by third-party solutions
like cert-manager, the route would not be successfully created due to the
dependency. This route will be rejected by the API server with an

> `FieldValueNotFound` error and will contain the reason as `referenced secret not present`.

Reviewer: This detail is not added in the implementation details for the api server validation point.

Author: Good catch, this scenario is not valid anymore since we should be allowed to create a route when the secret does not yet exist but the permission to read is created. This validation will only be part of the router.

management solutions like cert-manager.

### User Stories

- As an end user of Route API, I want to be able to provide a tls secret reference on
  OpenShift Routes so that I can integrate with third-party certificate management solution

- As an OpenShift cluster administrator, I want to use third-party solutions like cert-manager
  for certificate management of user workloads on OpenShift so that no manual process is
  required to renew expired certificates.

- As an end user of Route API, I want OpenShift Routes to support both manual and managed
  mode of operation for certification management so that I can switch between manual certificate
  management and third-party certificate management.

- As an OpenShift engineer, I want to be able to update Route API so that I can integrate
  OpenShift Routes with third-party certificate management solutions like cert-manager.

- An an OpenShift engineer, I want to be able to watch routes and process routes having valid
  annotations so that the certificate data is injected into the route CR.

- As an OpenShift engineer, I want to be able to run e2e tests as part of openshift/origin

>   so that this feature is frequently tests by the CI.

Reviewer 3: ```suggestion
  so that testcases are executed to signal feature health by CI executions.
```

  cert/key pair generated, the secret-injector controller copies the certificate data
  into the route at `.spec.tls.certificate` and `.spec.tls.key`.

**When an Ingress is used to expose user workload**

- The end user must have generated the serving certificate as a pre requisite
  using third-party systems like cert-manager.
- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
  where the Ingress is going to be created.
- To expose a user workload, he would need to create a new Ingress with the generated secret
  referenced in `.spec.tls.secretName`.
- If the secret CR that is referenced exists and has a successfully generated
  cert/key pair, the ingress-to-route controller copies this certificate data
  into the route at `.spec.tls.certificate` and `.spec.tls.key`.

_Note_: The same workflow would apply to updating an existing Route to integrate
with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate`
and `.spec.tls.key` in the route CR since the annotation gets a higher preference.

> #### Certificate Life cycle

Reviewer 3: It'd be good to detail the behavioral variations when ingress is configured with a secret name (not managed by cert-manager), route has annotation for secret.

Post integrating with third-party solutions like cert-manager, the end user can also
disable automatic certificate management. This can be done in 2 ways,

- The end user can delete the annotation on the Route.
- The end user deletes the secret that was associated with the Certificate CR.

**Deleting the annotation on the Route**

> The end user can delete the annotation on the Route, this will result

Reviewer 3: To be more clear, does it mean "When user deletes the annotation on the Route, default certificates generated by the cluster-ingress-operator are used and data at`.spec.tls` on Route object is cleared"
 ?

Author: Yes, I've updated to be more clear.

in abondoning the cert/key pair that was generated via cert-manager and resort to
using the default certificates that were initially generated by the cluster-ingress-operator
i.e. clearing the `.spec.tls` on Route.

> **Deleting the generated secret that contains the serving cert/key pair**

Reviewer 3: 1. Is the secret deletion behavior similar in case of Ingress? 
2. What'd be the usecase here for user to delete secrets? In case of compromised secrets, it'd be more useful for user to manual override by removing annotation or update certificate chain and keep the same secret. The secret is believed to be bearing the new certificates. 

In case certificate management is removed, it is better to expect user intervention to update the annotation as it'd an expected behavior for user.

Author: > Is the secret deletion behavior similar in case of Ingress?

The ingress-to-route controller, doesn't create a route when the secret is not found. Additional details are [here](https://github.com/openshift/enhancements/pull/1307#discussion_r1073254708).

> What'd be the usecase here for user to delete secrets?

Mostly just covering all bases (what ifs and other unintended actions).

When the end user deletes the secret containing the certificate data, the `.spec.tls` on
Route will be cleared. This will result in the route serving default certificates. In such
a scenario the annotation that was added by the user will be automatically deleted by the
secret-injector controller. This is only in the case where the user is directly creating

> routes instead of relying on the Ingresses.

Reviewer 3: Would it better to add user override for tls scenario ?

Author: This section is updated based on discussion [here](https://github.com/openshift/enhancements/pull/1307#discussion_r1081515918).

#### Variation [optional]

N.A

### API Extensions

A new annotation of type `string` to be introduced as part of the integration
to support third-party certificate management systems,

```yaml
annotations:

>   route.openshift.io/tls-secret-name: <secret-name>

Reviewer 4: I haven't reviewed this enhancement in detail, but a quick look at the API extensions section, I would lean towards using a new API field over the annotation approach. Creating annotation APIs is not discoverable and does not allow for validation or versioning, they are effectively, free form and guess work.

The `CertificateReference` in the `RouteSpec` is the way our API conventions would suggest you go

Author: So in this context the alternative proposal of `CertificateReference` would essentially point to a third party CRD like [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CRD provided by the `cert-manager`.

If what you are suggesting is that the `CertificateReference` (basically a `secretRef`/`secretName`) would just point to Secret of tls type, the route `.spec` would end up have multiple fields are related to the certificate data but they would be a oneoff (`.spec.tls.certificate`, `.spec.tls.key` OR `.spec.tls.certificateReference`). Is this what you are suggesting?

Author: The annotation was also kind of the preference of NE team initially. @alebedev87 and @Miciah Since we had put a pin on this until API review, do you think we can switch to a field (secretName) instead of an annotation?  

Another instance of this is seen in the gateway API as well as [CertificateRef](https://github.com/kubernetes-sigs/gateway-api/blob/df03bd58002fcf01ee3ffeba97edc1f0c67fb386/apis/v1beta1/gateway_types.go#L334) (although it is a cross namespace secret which we don't support)

Author: I believe there is another caveat to making this a field. Currently the router reads the certificate data from `.spec.tls.certificate` and `.spec.tls.key`, and it was a deliberate choice not read this directly from a secret (this reduces permissions required by the router and reduces attach surface on the edge). 

If we introduce a field, we still won't be able to remove `.spec.tls.certificate` and `.spec.tls.key`, so the logic would be something like, the secret is read and then copied onto the other 2 fields. Essentially one field helps in populating the other 2 field and this kind of seems weird (an annotation seems better suited to kind of configure this behaviour).

Reviewer 5: > Is having secret data embedded in this object not a security risk in itself? At least with a secret object users can introduce KMS encryption to protect the data, they don't have that option with data stored in a CRD. 

I hadn't heard of KMS before now.  Is KMS encryption incompatible with aggregated API servers?  The KMS documentation specifically calls out CRDs as supported in Kubernetes 1.26, but the documentation doesn't mention aggregated API servers such as openshift-apiserver, which handles the Route API in OpenShift: https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/#encrypting-your-data-with-the-kms-provider

(By the way, I've asked various people what makes secrets _actually_ more secret than other resources that are restricted by RBAC.  This might be the first substantive answer to that question that I've seen.  :tada:!)

> I would suggest revisiting that decision. Out of interest, were the pros and cons of reading secrets directly documented somewhere? Perhaps an arch discussion or enhancement can help give us the background?

The decision was made many years ago.  I don't know where it was written down, but it's basically this:  The openshift-router and haproxy processes run in the same pod, the haproxy process receives connections from untrusted clients and forwards connections to untrusted backend server pods, so it should be kept far away from anything particularly privileged, and so we don't want openshift-router to be privileged enough to read secrets.  

In short, the design that Thejas is proposing (and the design of the existing ingress-to-route controller) provides security through separation of concerns: one pod (with the secret-injector and ingress-to-route controllers) reads secrets and updates routes, and a different pod (the router) reads routes and handles network connections.  

The existence of KMS (which apparently can make some resources more secure than other resources) and the ability to do name-scoped watches on RBAC-restricted resources (which I learned about today) change the pro/con balance.  However, I'm wary of re-architecting openshift-router to establish and manage potentially hundreds of watches on secrets in the short time we have to implement this EP.

Reviewer 4: I feel like I'm having parallel conversations about this, but don't see the other thread 🤔 

I suspect you're right that the KMS integration isn't going to work with an aggregated API server, IIUC the requests for aggregated APIs are only proxied by the primary API server, so it wouldn't transform and implement the KMS encryption. That is a definite advantage to using a secret over an aggregated API, but you're right, we could also KMS encrypt a CRD.

IMO the main thing that makes secrets special is education, it's what users expect to be the most secure, because that's what the kube community has told them for a long time. Secrets were the first to have at least some encryption before going into storage, we now hacked that for aggregated APIs and CRDs too. Secrets are the most obvious target for KMS, but you can add CRDs there too. Ask the average user what to encrypt and I bet the first thing they say is a Secret 😉 

> In short, the design that Thejas is proposing (and the design of the existing ingress-to-route controller) provides security through separation of concerns: one pod (with the secret-injector and ingress-to-route controllers) reads secrets and updates routes, and a different pod (the router) reads routes and handles network connections.

That's a fair argument and thanks for the history

> However, I'm wary of re-architecting openshift-router to establish and manage potentially hundreds of watches on secrets in the short time we have to implement this EP.

I'm curious what the time pressure is here? My take on what's happening in OpenShift right now is that we are supposed to be focusing on quality and that deadlines for features and hitting specific releases shouldn't be our primary concern, if we think there's a better way to do the feature or rushing it will cause issues down the line, we should be able to say no, this isn't going to make 4.13. If missing that deadline means we can't ship 4.13, that's probably a different conversation but I would encourage you to think about the balance of a (arbitrary?) deadline vs getting the feature right

Author: Hey @JoelSpeed,
There were parallel discussions on this and in light of the recent discussions and implementation suggestions, we've decided to pick this up for the next release and not for 4.13 GA. So i'm putting this on hold for now and will pick this later (post 4.13 GA).

type RouteCertificateConditionType string

// RouteCertificateCondition contains details of the certificate being served by routers associated with
// this route.
type RouteCertificateCondition struct {
	// Type is the type of the condition.
    // Possible values include Default, Custom and Managed.
	Type RouteCertificateConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RouteCertificateConditionType"`
	// Status is the status of the condition.
	// Can be True, False, Unknown.
	Status corev1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
	// (brief) reason for the condition's last transition, and is usually a machine and human
	// readable constant
	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
	// Human readable message indicating details about last transition.
	Message string `json:"message,omitempty" protobuf:"bytes,4,opt,name=message"`
	// RFC 3339 date and time when this condition last transitioned
	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,5,opt,name=lastTransitionTime"`

> }

Reviewer 4: I wonder if you're overcomplicating this with a condition? How much of this data do you actually need?

Seems to me like the controller could just denote its certificate status with an enum? 
```
certificateType: Default | Custom | Managed
```

Author: Hmm, that is a good suggestion. All three types were mutually exclusive so an Enum could work. Is there some sort of convention for `.status`, because I don't see a lot of API's deviating to have something new only in the `.status` (something not replicated from `.spec`). 

Apart from using `Reason`/`Message` to indicate why said type was evaluated by the controller and `status` to indicate the type in use currently no other fields are required.

Author: I think showing the user the `Message`, briefly explaining the "why" about the certificate type is helpful.

Author: I think we can switch to using `metav1.Condition`. The message for example would say something like the following,
```
// update status to Managed
cond := metav1.Condition{
    Type: ManagedCertificate,
    Status: metav1.ConditionTrue,
    Message: "TLS secret reference annotation is provided, certificate data is managed by an external entity.",
}

// update status to Default
// update message as defaulting due to secret not found
cond := metav1.Condition{
    Type: DefaultCertificate,
    Status: metav1.ConditionTrue,
    Message: "TLS secret reference annotation present but associated secret is missing, using default generated certificates.",
}

// update status to Custom
cond := metav1.Condition{
    Type: CustomCertificate,
    Status: metav1.ConditionTrue,
    Message: "TLS secret reference annotation missing, using custom certificate data since provided in .spec",
}
```
The important one here is the 2nd example as this indicates the scenario discussed [here](https://github.com/openshift/enhancements/pull/1307/files#r1091307533).

Reviewer 4: That makes sense, lets go for `metav1.Condition` in this part then and keep it as in the examples you've provided

    // is being served on the route.
    ManagedCertificate RouteCertificateConditionType = "Managed"
)

```

#### Variation

##### Alternative to the tls-secret-reference annotation

**_Note_**: This idea has been dropped in favor of a secret reference through an annotation.

The alternative to the annotation was to provide a new API field in the Route API,
through which the user could provide the certificate reference.

```go
type CertificateReference struct {
	// certificate resource name.
	// +optional
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
}

Type RouteSpec struct {
	// ...

	CertificateRef CertificateReference `json:"certificateRef,omitempty" protobuf:"bytes,5,opt,name=certificateRef"`
}
```

The reasoning for the field was mainly for future proofing the API to handle integrations
with other third-party certificate management systems. But this API introduces inconsistency
between how Ingresses are supported by cert-manager via `.spec.tls.secretName` and when using
Routes, the user would have to provide the Certificate CR reference.

This also adds coupling with external third-party API(s) and limits future enhancement to be

> a more generic solution for integrating with third-party certificate management systems.

Reviewer 4: Does it? Are we not just saying that any certificate that can be provided by a secret can be used? That seem like it has no tighter coupling than the annotation API above

I actually don't see how this comment cannot apply to both the secret reference and annotation APIs, they both carry the same data, just one has a schema and the other doesn't

Author: Clarified [here](https://github.com/openshift/enhancements/pull/1307#discussion_r1085455475).

##### Alternative to new status condition types

> An alternative to introducing 3 new condition types, would be to use `metadata.managedFields`

Author: Hey @JoelSpeed, I would like to hear your thoughts on this kind of use-case for `managedFields`.

Reviewer 4: Can you expand on `are always managed by the controller as long as the annotation is present.`?

Are you suggesting that the controller would review the managed fields, and, if it is the managed of these (or there is no manager?) then it would continue to update them, but, if the managed fields suggests that there is another owner, then it would not manage them?

In general, overloading the managed fields strikes me as a bad idea if I've interpreted what you're trying to do correctly. It seems very implicit when APIs in Kube should be explicit and declarative

Author: Kind of yes, basically the manager attached to the fields (`.spec.tls.certificate` and `.spec.tls.key`) would determine who updated it and based on this being one of the pre-conditions the controller would take specific action.

Author: Agreed, `managedFields` wasn't designed for this use case.

### Implementation Details/Notes/Constraints [optional]

The new controller (secret-injector) will be responsible for watching routes and processing those
that have the valid annotation for `tls-secret-name`. This controller will also update
`.status` of the route to contain the latest information on the certificate in use.

Since secrets are a related resource to the secret-injector controller, the new controller
will also set up watches and re-sync routes in the same namespace as that of the secret (similar
to ingress-to-route controller).

The new controller will watch all routes and process all routes in order to update `.status.certificateCondiitions`.
The ones with the annotation will have additional logic to update `.spec.tls`.

In scenarios where the user deletes the annotation, it is the controller's
responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e.
if the status denotes `ManagedCertificate` and since the required annotation is
not present this results in the `.spec.tls` being cleared out and the `.status` updated
to `DefaultCertificate`.

> If the user deletes the secret associated with the Route wihtout deleting the `route.openshift.io/tls-secret-name`

Reviewer 3: ```suggestion
If the user deletes the secret associated with the Route without deleting the `route.openshift.io/tls-secret-name`
```

    - The role and rolebinding to provide this access must be provided by the user.
  - CEL validations and openshfit/library-go will enforce that both `.spec.tls.certificate` and `.spec.tls.externalCertificate`
    are not specified on the route at the same time. Since CEL validations are not run on Openshift API server, the same
    validation will be done as part of `ValidationRoute()`.

- New validation added to the API server as `ValidationHostExternalCertificate()`

  - Any route that is updated or created which has a non-empty `.spec.tls.externalCertificate`,
    will need require additional permission checks done as changing the certificate also affects
    `.spec.host` or `.spec.subdomain`. Meaning any user that is updating the certificate must also
    have `create` and `update` permission on the `custom-host` sub-resource.
  - Any user that does not have both these permissions will not be allowed to update/create routes
    that use `.spec.tls.externalCertificate`.
  - This validation is intentionally done only on the API server and not the router as well, this
    will result in not having this validation for events that are generated by the secret monitor

>     directly to the route controller in the router.

Reviewer: This is redundant to what you said in the drawbacks. We can omit it or reference the drawback section.

### Drawbacks

The user will need to manually create, provide and maintain the rbac required by the
router so that it can access secrets. This becomes tedious when users have
many routes.

The workaround for this is to document various levels of rbac that can be provided,

- Grant router service account access to secret by secret-name (explicit rbac)
- Grant router service account access to all secrets in a fixed namespace (implicit rbac)

The above variations need to be documented for the end user as part of OpenShift documentation.

#### Exception in Validations between API server and the router

The new `ValidationHostExternalCertificate()` is intentionally done only on the API server
and not the router as well, this will result in not having this validation for events that
are generated by the secret monitor directly to the route controller in the router. So if
a user who has the `create` and `update` permission on `custom-host` creates a route
that sets `.spec.tls.externalCertificate` the validation on the API server will pass and
the route is successfully created. Post creation of the route if the users permissions are

> revoked and the user edits the contents of the secret, this would successfully reload the

Reviewer: ```suggestion
revoked and the user edits the contents of the secret, the router would still be able to successfully reload the
```

**When an Ingress is used to expose user workload**

- The end user must have generated the serving certificate as a pre requisite
  using third-party systems like cert-manager.
- In cert-manager's case, the [Certificate](https://cert-manager.io/docs/usage/certificate/#creating-certificate-resources) CR must be created in the same namespace
  where the Ingress is going to be created.
- To expose a user workload, a new Ingress with the generated secret
  referenced in `.spec.tls.secretName` needs to be created.
- If the secret CR that is referenced exists and has a successfully generated
  cert/key pair, the ingress-to-route controller copies this certificate data
  into the route at `.spec.tls.certificate` and `.spec.tls.key`.

_Note_: The same workflow would apply to updating an existing Route to integrate
with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate`
and `.spec.tls.key` in the route CR since the annotation gets a higher preference.

#### Certificate Life cycle

Post integrating with third-party solutions like cert-manager, the end user can also
disable automatic certificate management. This can be done in 2 ways,

> - The end user can delete the annotation on the Route.

Author: Addressed, gone ahead with your suggestion.

- The end user can delete the secret-reference annotation on the Route.
- The end user deletes the secret that was associated with the Certificate CR.

**Deleting the secret-reference annotation on the Route**
The end user can delete the annotation on the Route, this will result
in clearing `.spec.tls.certificate` and `.spec.tls.key` and resort to

> using the default certificates that are generated by the cluster-ingress-operator.

Reviewer 5: I suppose we could have an annotation like `last-applied-configuration` to store the old data from `spec.tls.certificate` and `spec.tls.key`.  Would that be less confusing?

> using the default certificates that are generated by the cluster-ingress-operator.

Reviewer 4: Not sure how I feel about an annotation API, in general we steer away from them, but I guess this would be considered "internal" so it's probably ok? The idea of storing it to restore later seems beneficial though, +1 to that

When the end user deletes the secret containing the certificate data, the `.spec.tls` on
Route will be cleared. This will result in the route serving default certificates. In such
a scenario the annotation that was added by the user will be not deleted by the
secret-injector controller and since the secret mentioned is not present, the controller
will clear `.spec.tls.certificate` and `.spec.tls.key` and instead use the default
certificates that are generated by the cluster-ingress-operator. This is only in the case

> where the user is directly creating routes instead of relying on the Ingresses.

Author: This was left over from an earlier stage. I've removed this line.

#### Variation [optional]

N.A

### API Extensions

A new annotation of type `string` to be introduced as part of the integration
to support third-party certificate management systems,

```yaml
annotations:
  route.openshift.io/tls-secret-name: <secret-name>
```

_Note_: The default value would be N/A. The secret is required to be created
in the same namespace as that of the Route. The secret must be of type
`kubernetes.io/tls` and the tls.key and the tls.crt key must be provided in
the `data` (or `stringData`) field of the Secret configuration.

This annotation can be applied by the end user on Route. The controller that
will be introduced as part of this enhancement will be responsible for the
processing of this annotation on the Route.

The Route API will also be updated to denote certificate status under `.status`,

```go

// RouteStatus provides relevant info about the status of a route, including which routers
// acknowledge it.
type RouteStatus struct {
    // ...

    // CertificateConditions describes the type of certificate that is being served
    // by the router(s) associated with this route.

>     Certificate []RouteCertificateCondition `json:"certificate,omitempty" protobuf:"bytes,2,rep,name=certificate"`

Author: I'll leave this open until @JoelSpeed gets back on the API discussion [here](https://github.com/openshift/enhancements/pull/1307/files#r1082397811).

	// (brief) reason for the condition's last transition, and is usually a machine and human
	// readable constant
	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
	// Human readable message indicating details about last transition.
	Message string `json:"message,omitempty" protobuf:"bytes,4,opt,name=message"`
	// RFC 3339 date and time when this condition last transitioned
	LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,5,opt,name=lastTransitionTime"`
}

const (
    // DefaultCertificate denotes that the user has not provided
    // any custom certificate and the default generated certificate is being
    // served on the route.
    DefaultCertificate RouteCertificateConditionType = "Default"

    // CustomCertificate denotes that there is a custom certificate
    // being served on the route.
    CustomCertificate RouteCertificateConditionType = "Custom"

    // ManagedCertificate denotes that the annotation is in use to
    // inject a third-party managed secret containing the certificate that
    // is being served on the route.
    ManagedCertificate RouteCertificateConditionType = "Managed"

> )

Author: Updated as a goal,
>- Provide latest certificate type information on the Route.

- When transitioning from using managed certificates to manually managing custom certificates,
  the `.managedFields` will still depict `.spec.tls` as managed by the controller since the update
  operation done by the user is only on the `.metadata.annotations`.
- Also since `oc` defaults to `--server-side=false`, mixing CSA and SSA results in ambiguous `.managedFields`.

### Implementation Details/Notes/Constraints [optional]

The new controller (secret-injector) will be responsible for watching routes and processing those
that have the valid annotation for `tls-secret-name`. This controller will also update
`.status` of the route to contain the latest information on the certificate in use.

Since secrets are a related resource to the secret-injector controller, the new controller
will also set up watches and re-sync routes in the same namespace as that of the secret (similar
to ingress-to-route controller).

The new controller will watch all routes and process all routes in order to update
`.status.certificate`. The ones with the annotation will have additional
logic to update `.spec.tls.certificate` and `.spec.tls.key` based on the following pre-conditions,

> - Annotation is present and valid

Author: After making the change regarding the annotation name, "valid" doesn't really make sense. So I've removed the same.

> - Secret mentioned in the annotation exists and is valid

Author: Addressed.

In scenarios where the user deletes the annotation, it is the controller's

> responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e.

Author: I'll leave this open until @JoelSpeed gets back on the API discussion [here](https://github.com/openshift/enhancements/pull/1307/files#r1082397811).

> if the status denotes `ManagedCertificate` and since the required annotation is

Reviewer: I suppose that "the latest certificate status condition has to be `ManagedCertificate`" would be the exact description of the controller's logic here, no?

Author: Addressed.

out and the `.status` updated to `DefaultCertificate`.

If the user deletes the secret associated with the Route without deleting the
`route.openshift.io/tls-secret-name` annotation, the secret-injector controller will
clear out `.spec.tls.certificate` and `.spec.tls.key` since the following pre-conditions

> are met,

Reviewer 5: You could call it an error condition.  The user specified incomplete configuration (dangling cert reference), so the controller makes its best effort (use the default cert) and notifies the user of the issue via an event and status condition.  How is that opaque?  It's consistent with the behavior of the ingress-to-route controller (if you delete the secret that an ingress references, the ingress-to-route controller deletes any associated routes).

> are met,

Reviewer 4: (trying to remember what I was thinking 2 weeks ago...)

In terms of the route itself, what would a user see if they tried to access the endpoint? Possibly a certificate error? Possibly not? 

I think I was going down the line of, if the endpoint suddenly failed spectacularly, it would be obvious there was an error in the route, but that's probably not a great UX for something like a Route API, so your current behaviour from ingress-to-route is more pleasant in that regard

are met,

- Annotation is present and valid
- Secret mentioned in the annotation doesn't exist

> - Route `.status.certificate` has `ManagedCertificate`=`True`

Reviewer: Shouldn't this be the latest condition to be precise?

Author: Addressed.

  which results in the router using the default certificates that are generated and

>   the `.status` updated as well.

Reviewer: You were describing the pre-conditions here, why saying about the results? This seems to belong in the phrases just after.

Author: Corrected the indentation.

As a follow up to the above scenario when the secret-reference annotation added by the user is retained
by the secret-injector controller and because the referenced secret is not present,
the controller will publish an `Event` so that the end user is notified regarding this
switch to using the default generated certificates. The `.status` is also updated to `DefaultCertificate`.

When transitioning from using a managed certificate to manually managed certificate, the user
is expected to delete the secret-reference annotation and provide new certificate/key pair under `.spec.tls`.
If the previously used `.spec.tls.certificate` and `.spec.tls.key` is not replaced

> by the user, the secret-injector controller will clear `.spec.tls.certificate` and `.spec.tls.key`

Reviewer 4: How will the controller know if it was the value from the secret or the new values from the user? If this were a spec field referencing a secret, this would be unambiguous

2. How/where to implement e2e tests for [route-controller-manager](https://github.com/openshift/route-controller-manager)?
   [CLOSED]

**Answer**: The e2e tests will be added to openshift/origin.

### Test Plan

This enhancement will be tested in isolation of cert-manager-operator as part of
core OCP payload. This will also be tested with cert-manager-operator as part of the
operators e2e test suite.

1. Test Route API without interfacing with Ingresses

   a. Create a edge terminated route that is using default certificates.
   b. Update route with the secret-reference annotation and ensure the `.spec.tls` and `.status`
   is updated.
   c. Verify that the associated openshift routers are also serving the updated
   certificates and not the default certificates.
   d. Update certificate data and verify if the certificates served by the router
   are updated as well.
   d. Delete the annotation and ensure that the route `.spec.tls` and `.status`
   denotes the usage of default certificates. Also verify the associated
   openshift router is serving the default certificates.

Various other transitions, `Custom`->`Managed`-`Custom` will also be tested.

### Graduation Criteria

> This is a user facing change and will directly go to GA. This feature requires an

Reviewer 4: Why not techpreview? Is there no merit in developing this in a techpreview environment and enabling it later? Individual API fields can be added as TechPreview with our current tooling
